File: src/Systems/C/facade.c
#include <stdio.h>

static void authenticate(char *out, size_t size, const char *user) {
    snprintf(out, size, "auth(%s)", user);
}

static void reserve(char *out, size_t size, const char *sku) {
    snprintf(out, size, "reserve(%s)", sku);
}

static void charge(char *out, size_t size, int cents) {
    snprintf(out, size, "charge(%d)", cents);
}

static void checkout(char *out, size_t size, const char *user, const char *sku, int cents) {
    char auth[64];
    char inventory[64];
    char billing[64];
    authenticate(auth, sizeof auth, user);
    reserve(inventory, sizeof inventory, sku);
    charge(billing, sizeof billing, cents);
    snprintf(out, size, "checkout=%s>%s>%s", auth, inventory, billing);
}

int main(void) {
    char result[256];
    checkout(result, sizeof result, "alice", "SKU-42", 499);
    puts(result);
    return 0;
}

File: src/Historical/Fortran/facade.f90
module facade_services
  implicit none
contains
  function authenticate(user) result(text)
    character(len=*), intent(in) :: user
    character(len=:), allocatable :: text
    text = 'auth(' // trim(user) // ')'
  end function authenticate

  function reserve(sku) result(text)
    character(len=*), intent(in) :: sku
    character(len=:), allocatable :: text
    text = 'reserve(' // trim(sku) // ')'
  end function reserve

  function charge(cents) result(text)
    integer, intent(in) :: cents
    character(len=32) :: amount
    character(len=:), allocatable :: text
    write(amount, '(I0)') cents
    text = 'charge(' // trim(amount) // ')'
  end function charge

  function checkout(user, sku, cents) result(text)
    character(len=*), intent(in) :: user, sku
    integer, intent(in) :: cents
    character(len=:), allocatable :: text
    text = 'checkout=' // authenticate(user) // '>' // reserve(sku) // '>' // charge(cents)
  end function checkout
end module facade_services

program facade_example
  use facade_services
  implicit none
  print '(A)', checkout('alice', 'SKU-42', 499)
end program facade_example
